jena-arq/src/main/java/org/apache/jena/sparql/engine/QueryExecutionBase.java

```diff
@@ -242,6 +257,49 @@ public Model execConstruct(Model model)
         Template template = query.getConstructTemplate() ;
         return TemplateLib.calcTriples(template.getTriples(), queryIterator);
     }
+    
+    @Override
+    public Iterator<Quad> execConstructQuads()
+    {
+        checkNotClosed() ;
+        if ( ! query.isConstructType() )
+            throw new QueryExecException("Attempt to get a CONSTRUCT model from a "+labelForQuery(query)+" query") ;
+        // This causes there to be no PROJECT around the pattern.
+        // That in turn, exposes the initial bindings.  
+        if ( ! Syntax.syntaxARQ.equals( query.getSyntax() ) )
+        	throw new QueryExecException("Attempt to CONSTRUCT quads from a "+labelForQuery(query)+" query, which is not ARQ Syntax") ;
+        	
+        query.setQueryResultStar(true) ;
+
+        startQueryIterator() ;
+        
+        Template template = query.getConstructTemplate() ;
+        return TemplateLib.calcQuads(template.getQuads(), queryIterator);
+    }
+    
+    @Override
+    public Dataset execConstructDataset(){
+    	
+    	DatasetGraph graph = DatasetGraphFactory.createMem();
+    	
+        checkNotClosed() ;
+        try
+        {
+            Iterator<Quad> it = execConstructQuads();
+                        
+            while (it.hasNext())
+            {
+                Quad q = it.next();
+                graph.add(q);
```

Comment:
Given Java 8 this iterator loop could be just `execConstructQuads().forEachRemaining(graph::add);`